### PR TITLE
Typecheck let expressions

### PR DESCRIPTION
**Pull request by janmasrovira**
Progress for #1351.

This pr  includes:
- Let expressions for abstract and internal languages
- Termination, arity and type checking for let expressions
- Some other fixes

### CODE REVIEW

File: src/Juvix/Compiler/Abstract/Translation/FromConcrete.hs [hunk 1]
```diff
@@ -336,6 +336,39 @@ goExpression = \case
...
+                getClauses :: Sem r (NonEmpty Abstract.FunctionClause)
+                getClauses = do
+                  cls <-
+                    sequence
+                      [ goFunctionClause c | LetFunClause c <- toList cl, sig ^. sigName == c ^. clauseOwnerFunction
+                      ]
+                  case nonEmpty cls of
+                    Nothing ->
+                      throw
+                        ( ErrLacksFunctionClause (LacksFunctionClause sig)
```
Comment (paulcadman) [inline src/Juvix/Compiler/Abstract/Translation/FromConcrete.hs:368]: Add a negative test for this? (Unless there's already one I've missed).

File: src/Juvix/Compiler/Internal/Language.hs [hunk 2]
```diff
@@ -59,20 +65,26 @@ data FunctionDef = FunctionDef
     _funDefClauses :: NonEmpty FunctionClause,
     _funDefBuiltin :: Maybe BuiltinFunction
   }
+  deriving stock (Eq, Generic, Data)
+
+instance Hashable FunctionDef
```
Comment (paulcadman) [inline src/Juvix/Compiler/Internal/Language.hs:70]: Is this Hashable instance used?

File: src/Juvix/Compiler/Internal/Language.hs [hunk 3]
```diff
@@ -59,20 +65,26 @@ data FunctionDef = FunctionDef
...
 
 data FunctionClause = FunctionClause
   { _clauseName :: FunctionName,
     _clausePatterns :: [PatternArg],
     _clauseBody :: Expression
   }
+  deriving stock (Eq, Generic, Data)
+
+instance Hashable FunctionClause
```
Comment (paulcadman) [inline src/Juvix/Compiler/Internal/Language.hs:79]: Is this Hashable instance used?

File: src/Juvix/Compiler/Internal/Language.hs (hunk 2 above)
Comment (paulcadman) [reply src/Juvix/Compiler/Internal/Language.hs:70]: yes - for deriving Expression

File: src/Juvix/Compiler/Internal/Language.hs (hunk 3 above)
Comment (paulcadman) [reply src/Juvix/Compiler/Internal/Language.hs:79]: yes - for deriving Hashable Expression

**Review by paulcadman — APPROVED** [review]

File: src/Juvix/Compiler/Abstract/Translation/FromConcrete.hs (hunk 1 above)
Comment (janmasrovira) [reply src/Juvix/Compiler/Abstract/Translation/FromConcrete.hs:368]: done in f8069ba7

**Review by paulcadman — DISMISSED** [review]

**Review by paulcadman — APPROVED** [review]

**Pull request merged**